events/events.go

```diff
@@ -0,0 +1,197 @@
+package events
+
+import (
+	"errors"
+	"sync"
+)
+
+var errClosed = errors.New("publisher is closed")
+
+// GetFn returns the event
+type GetFn func() interface{}
+
+// Event is an event
+type Event interface {
+	// OnGet sets the function to be executed when Get() is called
+	OnGet(GetFn)
```

Comment:
Feels a bit strange that we have an `OnGet` method on the Event interface. Seems this is only used to register the getter function when a subscriber is created? If so can we do something like below and get rid of the `OnGet` method?
```
func (p *publisher) GetAndSubscribe() (interface{}, Subscriber, error) {
  subscriber, _ := p.newSubscriber()  // ignore error handling here for simplicity
  return p.Get(), subscriber, nil
}

func (p *publisher) newSubscriber() (Subscriber, error) {
  p.Lock()
  if p.closed {
    p.Unlock()
    return errPublisherClosed
  }
  notificationCh := make(chan struct{}, 1)
  subscriber := newSubscriber(notificationCh, p.Get)  
  p.subscribers = append(p.subscribers, subscriber)
  p.Unlock()
  return subscriber, nil
}

type subscriber struct {
  ch <-chan struct{}
  getFn GetFn
  ...  // other fields
}

func newSubscriber(ch <-chan struct{}, getFn GetFn) Subscriber {
  return subscriber{ch: ch, getFn: getFn}
}

func (s *subscriber) Get() interface{} {
  return s.getFn()
}
```